### ISSUE TITLE
gitlab repository webhook not working

### ISSUE THREAD

**Issue by thudry** [ovh/cds#4045]
I was unable to add a repository webhook, we use gitlab 11.7.5.
It seem gitlab return a 422 to cds.

cds-api logs:
```
2019-03-18 15:29:37 [WARN] pprofLabel>recoverWrap>Handle>putWorkflowHandler>HookRegistration>createVCSConfiguration>CreateHook>doJSONRequest: internal server error (caused by: putWorkflowHandler> HookRegistratio
n: Cannot update vcs configuration: Cannot create hook on repository: {ID: Name: Events:[] Method:POST URL:https://cds.****.com/webhooks/webhook/d1cc1a0b-2166-4b76-8218-d85fb2ee1329 ContentType: Body: Disabl
e:false InsecureSSL:false Workflow:true}: unable to create hook on repository thudry/test-chart from gitlab: pprofLabel>recoverWrap>Handle>putWorkflowHandler>HookRegistration>createVCSConfiguration>CreateHook>do
JSONRequest>DoJSONRequest>doJSONRequest: internal server error (caused by: Unable to perform request on service corp-cds-cds-vcs (vcs))) error_uuid=a3d051d0-4992-11e9-b432-0a5864602f86 method=PUT request_uri=/pr
oject/TEST/workflows/bbp stack_trace=pprofLabel>recoverWrap>Handle>putWorkflowHandler>HookRegistration>createVCSConfiguration>CreateHook>doJSONRequest>DoJSONRequest>doJSONRequest: internal server error (caused b
y: Unable to perform request on service corp-cds-cds-vcs (vcs))                                                                                                                                                    
github.com/ovh/cds/sdk.WrapError                                                                                                                                                                                   
        /go/src/github.com/ovh/cds/sdk/error.go:666                                                                                                                                                                
github.com/ovh/cds/engine/api/repositoriesmanager.(*vcsClient).doJSONRequest                                                                                                                                       
        /go/src/github.com/ovh/cds/engine/api/repositoriesmanager/repositories_manager.go:211                                                                                                                      
github.com/ovh/cds/engine/api/repositoriesmanager.(*vcsClient).CreateHook                                                                                                                                          
        /go/src/github.com/ovh/cds/engine/api/repositoriesmanager/repositories_manager.go:392                                                                                                                      
github.com/ovh/cds/engine/api/workflow.createVCSConfiguration                                                                                                                                                      
        /go/src/github.com/ovh/cds/engine/api/workflow/hook.go:232                                                                                                                                                 
github.com/ovh/cds/engine/api/workflow.HookRegistration                                                                                                                                                            
        /go/src/github.com/ovh/cds/engine/api/workflow/hook.go:147                                                                                                                                                 
github.com/ovh/cds/engine/api.(*API).putWorkflowHandler.func1                                                                                                                                                      
        /go/src/github.com/ovh/cds/engine/api/workflow.go:459                                                                                                                                                      
github.com/ovh/cds/engine/api.(*Router).Handle.func1                                                                                                                                                               
        /go/src/github.com/ovh/cds/engine/api/router.go:245                                                                                                                                                        
net/http.HandlerFunc.ServeHTTP                                                                                                                                                                                     
        /usr/local/go/src/net/http/server.go:1947                                                                                                                                                                  
github.com/ovh/cds/engine/api.(*Router).recoverWrap.func1                                                                                                                                                          
        /go/src/github.com/ovh/cds/engine/api/router.go:152                                                                                                                                                        
net/http.HandlerFunc.ServeHTTP                                                                                                                                                                                     
        /usr/local/go/src/net/http/server.go:1947                                                                                                                                                                  
github.com/ovh/cds/vendor/github.com/gorilla/handlers.CompressHandlerLevel.func1                                                                                                                                   
        /go/src/github.com/ovh/cds/vendor/github.com/gorilla/handlers/compress.go:143                                                                                                                              
net/http.HandlerFunc.ServeHTTP                                                                                                                                                                                     
        /usr/local/go/src/net/http/server.go:1947                                                                                                                                                                  
net/http.(Handler).ServeHTTP-fm                                                                                                                                                                                    
        /go/src/github.com/ovh/cds/engine/api/router.go:94                                                                                                                                                         
github.com/ovh/cds/engine/api.(*Router).pprofLabel.func1                                                                                                                                                           
        /go/src/github.com/ovh/cds/engine/api/router.go:89                                                                                                                                                         
net/http.HandlerFunc.ServeHTTP                                                                                                                                                                                     
        /usr/local/go/src/net/http/server.go:1947
github.com/ovh/cds/vendor/github.com/gorilla/mux.(*Router).ServeHTTP
        /go/src/github.com/ovh/cds/vendor/github.com/gorilla/mux/mux.go:114
net/http.serverHandler.ServeHTTP                                           
        /usr/local/go/src/net/http/server.go:2694                                                                                                                                                                  net/http.(*conn).serve                                                                                                                                                                                                     /usr/local/go/src/net/http/server.go:1830                                                                                                                                                                 
runtime.goexit                  
        /usr/local/go/src/runtime/asm_amd64.s:2361 
unable to create hook on repository thudry/test-chart from gitlab
Cannot create hook on repository: {ID: Name: Events:[] Method:POST URL:https://cds.****.com/webhooks/webhook/d1cc1a0b-2166-4b76-8218-d85fb2ee1329 ContentType: Body: Disable:false InsecureSSL:false Workflow:$
rue}                                                
Cannot update vcs configuration                                   
putWorkflowHandler> HookRegistration status=500 
```

cds-vcs logs:
```
2019-03-18 15:29:37 [WARN] pprofLabel>recoverWrap>Handle>postHookHandler>CreateHook: internal server error (caused by: CreateHook gitlab thudry/test-chart: AddProjectHook: POST https://gitlab.****.com/api/v4/projects/thudry/test-chart/hooks: 422 {error: Invalid url given}) error_uuid=a3cfdc1c-4992-11e9-ab2c-0a58646013de method=POST request_uri=/vcs/gitlab/repos/thudry/test-chart/hooks stack_trace=POST https://gitlab.****.com/api/v4/projects/thudry/test-chart/hooks: 422 {error: Invalid url given}
AddProjectHook
github.com/ovh/cds/sdk.WrapError
        /go/src/github.com/ovh/cds/sdk/error.go:673
github.com/ovh/cds/engine/vcs/gitlab.(*gitlabClient).CreateHook
        /go/src/github.com/ovh/cds/engine/vcs/gitlab/client_hook.go:50
github.com/ovh/cds/engine/vcs.(*Service).postHookHandler.func1
        /go/src/github.com/ovh/cds/engine/vcs/vcs_handlers.go:836
github.com/ovh/cds/engine/api.(*Router).Handle.func1
        /go/src/github.com/ovh/cds/engine/api/router.go:245
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:1947
github.com/ovh/cds/engine/api.(*Router).recoverWrap.func1
        /go/src/github.com/ovh/cds/engine/api/router.go:152
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:1947
github.com/ovh/cds/vendor/github.com/gorilla/handlers.CompressHandlerLevel.func1
        /go/src/github.com/ovh/cds/vendor/github.com/gorilla/handlers/compress.go:143
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:1947
net/http.(Handler).ServeHTTP-fm
        /go/src/github.com/ovh/cds/engine/api/router.go:94
github.com/ovh/cds/engine/api.(*Router).pprofLabel.func1
        /go/src/github.com/ovh/cds/engine/api/router.go:89
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:1947
github.com/ovh/cds/vendor/github.com/gorilla/mux.(*Router).ServeHTTP
        /go/src/github.com/ovh/cds/vendor/github.com/gorilla/mux/mux.go:114
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2694
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1830
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:2361
CreateHook gitlab thudry/test-chart status=500

```

**Comment by richardlt** [comment]
Hi, the url that is given to gitlab looks good (https://cds.****.com/webhooks/webhook/d1cc1a0b-2166-4b76-8218-d85fb2ee1329), and use https.
Is the certificate a valid and not self signed one ?

**Comment by thudry** [comment]
The certificate is not self signed and is valid.

**Comment by thudry** [comment]
Here are the related gitlab logs

```
{"time":"2019-03-18T16:20:07.704Z","severity":"INFO","duration":48.9,"db":4.47,"view":44.43,"status":422,"method":"POST","path":"/api/v4/projects/thudry%2Ftest-chart/hooks","params":[{"key":"url","value":"https:
//cds.***.com/webhooks/webhook/9d168747-cad4-4bb4-ae15-231a44539466"},{"key":"push_events","value":null},{"key":"merge_requests_events","value":null},{"key":"tag_push_events","value":null},{"key":"enable_ss
l_verification","value":null}],"host":"gitlab.***.com","ip":"100.96.16.0, 100.96.16.0","ua":"go-gitlab/0.2.0","route":"/api/:version/projects/:id/hooks","user_id":55,"username":"thudry","queue_duration":8.5
3,"gitaly_calls":0,"correlation_id":"Im64FXSVfu8"}
```

I wonder if this could be related with https://github.com/ovh/cds/issues/3964#issuecomment-466820833

**Comment by richardlt** [comment]
I don't think this is related, you are not using subgroup thudry (your project), test-chart (the repo).
Did you try to setup manually the hook in Gitlab ui (your repo settings/integrations) to see if Gitlab returns more info ?
It works for Gitlab cloud instance, looks like you are using a private gitlab instance. Is this instance compatible with /api/v4/ api calls ?

**Comment by thudry** [comment]
Yes we are indeed using a private instance.
I found the issue when I tried to add the webhook manually, it turns out that by default gitlab restrict outbound requests to the local network.
I followed https://docs.gitlab.com/ee/security/webhooks.html and I was able to create the webhook from csd.
It might be worth to add a note for on premise instances on https://ovh.github.io/cds/hosting/repositories-manager/gitlab/

Thanks for the help ! :)
